src/compiler/checker.ts

```diff
@@ -21833,7 +21836,11 @@ namespace ts {
             if (isBindingPattern(node.name)) {
                 // Don't validate for-in initializer as it is already an error
                 if (node.initializer && node.parent.parent.kind !== SyntaxKind.ForInStatement) {
-                    checkTypeAssignableTo(checkExpressionCached(node.initializer), getWidenedTypeForVariableLikeDeclaration(node), node, /*headMessage*/ undefined);
+                    let initializerType = checkExpressionCached(node.initializer);
+                    if (strictNullChecks && node.name.elements.length === 0) {
+                        initializerType = checkNonNullType(initializerType, node);
+                    }
+                    checkTypeAssignableTo(initializerType, getWidenedTypeForVariableLikeDeclaration(node), node, /*headMessage*/ undefined);

```

Comment:
Is checkTypeAssignableTo needed if there aren't any elements? Seems like it could in an `else`, and then `initializerType` wouldn't need to be updated in the `if` block.

Reply:
Good point. I did consider whether it would be possible to just return (or use an else) immediately after the null check. I was just unsure if `checkTypeAssignableTo` performs some side-effects (like caching in `checkTypeRelatedTo`) that might give different results if skipped.

Making that change doesn't break any tests so from what I can tell I think this would be ok. I'm still learning the details of the internals so I may be missing something.

Reply:
You're right, the cache might be filled earlier this way, but there's no point in doing work that isn't needed.

You can't return because the code still needs to `checkParameterInitalizer`, although maybe that's redundant with the new call to `checkExpressionCached`.

Reply:
I don't think `checkParameterInitalizer` gets called via `checkExpressionCached`, so unless it does some equivalent work, I need to move the call after the if-then-else, rather than in the else branch. My most recent commit got this wrong - will fix now.